### ISSUE TITLE
QtWebengine GPU usage causes blank Help Browser on some systems

### ISSUE THREAD

**Issue by zacksettel** [supercollider/supercollider#4136]
Bug description:

(using git checked out version 3.10, build on  MacOS 10.13.6 ) 

The Help window remains blank

Steps to reproduce:
1.  Search for Help on a topic  (Cmd-D on highlighted item, or any other way to get help)
2.
3.

Actual result:

Nothing appears in help window
Expected result:

information appears in help window

**Comment by patrickdupuis** [comment]
I don't have a macOS system to test at hand. Can you try running SuperCollider from within your build folder? Should be `build/Install/SuperCollider/SuperCollider.app`.

**Comment by zacksettel** [comment]
Same behaviour as when run from /Applications

**Comment by patrickdupuis** [comment]
This is possibly a duplicate of #4066. Can you try seeing if you get the same results as what is described there?

**Comment by zacksettel** [comment]
On my 3.10-beta, on OSX 10.13.6, I ran directly from the executable, as suggested above, and got the errors below.

dyld: Library not loaded: @loader_path/../../../../../../../QtQuick.framework/Versions/5/QtQuick
Referenced from: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Helpers/QtWebEngineProcess.app/Contents/MacOS/QtWebEngineProcess
Reason: image not found
[33480:51715:1024/175345.763033:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
[33480:51715:1024/175345.763235:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
[33480:51715:1024/175345.763295:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
dyld: Library not loaded: @loader_path/../../../../../../../QtQuick.framework/Versions/5/QtQuick
Referenced from: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Helpers/QtWebEngineProcess.app/Contents/MacOS/QtWebEngineProcess
Reason: image not found
[33480:51715:1024/175346.524974:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
[33480:51715:1024/175346.542240:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
[33480:51715:1024/175346.542419:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
[33480:51715:1024/175346.542483:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
dyld: Library not loaded: @loader_path/../../../../../../../QtQuick.framework/Versions/5/QtQuick
Referenced from: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Helpers/QtWebEngineProcess.app/Contents/MacOS/QtWebEngineProcess
Reason: image not found
[33480:51715:1024/175346.563731:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
[33480:51715:1024/175346.563800:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service
[33480:51715:1024/175346.563845:ERROR:service_manager_context.cc(252)] Attempting to run unsupported native service: /Applications/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/content_renderer.service

**Comment by zacksettel** [comment]
One more observation: 3.10-beta will configure (cmake) without error with the "qt5" package installed (via brew), but when running cmake for 3.9 (with the same options), cmake will fail, complaining about missing items (e.g. QtWebKit)....

Thus, 3.10-beta will configure and build, and run. But the help window will be blank.

**Comment by patrickdupuis** [comment]
The move from QtWebkit to QtWebengine is one of the major changes in `3.10`. If you install webkit, you'll be able to build `3.9.3`.

**Comment by zacksettel** [comment]
Build dir is not on a shared drive.

**Comment by zacksettel** [comment]
Brew does not have anything when I search for webkit, but in macports, there seem to be different flavours when I search for webkit.

Any recommendations about which one to try?


qt5-qtwebkit @5.9.1_2 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt5-qtwebkit-docs @5.9.1_2 (aqua)
    Documentation for the port qt5-qtwebkit

qt5-qtwebkit-examples @5.9.1_2 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt55-qtwebkit @5.5.1_4 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt55-qtwebkit-docs @5.5.1_4 (aqua)
    Documentation for the port qt55-qtwebkit

qt55-qtwebkit-examples @5.5.1_1 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt56-qtwebkit @5.6.3_1 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt56-qtwebkit-docs @5.6.3_1 (aqua)
    Documentation for the port qt56-qtwebkit

qt56-qtwebkit-examples @5.6.3 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt57-qtwebkit @5.7.1_1 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt57-qtwebkit-docs @5.7.1_1 (aqua)
    Documentation for the port qt57-qtwebkit

qt57-qtwebkit-examples @5.7.1 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt58-qtwebkit @5.8.0_1 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt58-qtwebkit-docs @5.8.0_1 (aqua)
    Documentation for the port qt58-qtwebkit

qt58-qtwebkit-examples @5.8.0 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt59-qtwebkit @5.9.1_1 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

qt59-qtwebkit-docs @5.9.1_1 (aqua)
    Documentation for the port qt59-qtwebkit

qt59-qtwebkit-examples @5.9.1 (aqua)
    Tools and Module(s) for Qt Tool Kit 5

surf @2.0 (www)
    simple webkit2-gtk-based browser

webkit-gtk @2.4.11_2 (www, gnome)
    Apple's WebKit HTML rendering library for GTK+

webkit-gtk-2.0 @2.0.4_9 (www, gnome)
    Apple's WebKit HTML rendering library for GTK+

webkit-gtk3 @2.4.11_2 (www, gnome)
    Apple's WebKit HTML rendering library for GTK+

webkit-gtk3-2.0 @2.0.4_9 (www, gnome)
    Apple's WebKit HTML rendering library for GTK+

webkit-sharp @0.3 (devel, www, gnome)
    C#/CLI bindings to WebKit/Gtk+

webkit2-gtk @2.20.5 (www, gnome)
    Apple's WebKit2 HTML rendering library for GTK+3 (with optional support for GTK+2 plugins)

webkit2-gtk-devel @2.21.91 (www, gnome)
    Apple's WebKit2 HTML rendering library for GTK+3 (with optional support for GTK+2 plugins)

**Comment by patrickdupuis** [comment]
The README_MACOS from 3.9 states:

>>git, cmake, libsndfile, readline, and qt5.5, installed via homebrew: brew install git cmake readline qt55
>>
>>Note: SuperCollider depends on Qt5WebKit, which was dropped from Qt in 5.6. So make sure you request Qt5.5 (qt55) and not the latest version (qt5). Use brew info to see the version of a package, and see brew switch or brew pin if you want to use Qt5.5 as your default version.

**Comment by patrickdupuis** [comment]
Some system info: 
Mac Pro (late 2013)
macOS High Sierra 10.13.6
Intel Xeon E5 quad-core 3.7ghz
12gb RAM

When using the latest s3 build, the SC post window says:
```
`Compiling directory ‘/private/var/folders/3h/r83s_xps6_vc8qfdd5848ttc0000gq/T/AppTranslocation/8CD0E186-5136-486D-8893-E8050FEB75F5/d/SuperCollider.app/Contents/Resources/SCClassLibrary’
AVDCreateGPUAccelerator: Error loading GPU renderer`
```

The TERMINAL says:
```
AVDCreateGPUAccelerator: Error loading GPU renderer
[987:51459:1025/092455.208555:ERROR:gl_context_cgl.cc(136)] Error creating context.
[987:51459:1025/092455.208611:ERROR:gles2_command_buffer_stub.cc(263)] ContextResult::kFatalFailure: Failed to create shared context for virtualization.
[987:51459:1025/092455.210211:ERROR:gl_context_cgl.cc(136)] Error creating context.
[987:51459:1025/092455.210236:ERROR:gles2_command_buffer_stub.cc(263)] ContextResult::kFatalFailure: Failed to create shared context for virtualization.
[987:51459:1025/092455.210955:ERROR:gl_context_cgl.cc(136)] Error creating context.
```
The HelpBrowser remains blank as stated in the initial report.

**Comment by patrickdupuis** [comment]
Related Qt forum posts:
https://forum.qt.io/topic/86831/avdcreategpuaccelerator-error-loading-gpu-renderer-on-macpro-qt-5-6-x-5-9-2

**Comment by patrickdupuis** [comment]
The OP reported elsewhere that this fixes the problem for them:
```
cd /Applications/SuperCollider/SuperCollider.app/Contents/MacOS/
./SuperCollider --disable-gpu
```
Here's the QtWebengine debugging info that led to this discovery:
https://doc.qt.io/qt-5.11/qtwebengine-debugging.html#using-command-line-arguments

**Comment by ahihi** [comment]
still a problem with 3.11.0 and macOS 10.13.6.

the workaround works, so i wrote a script for patching the .app to always run with `--disable-gpu`:

```bash
#!/bin/bash
APP=$1

if [ ! -n "$APP" ]; then
    echo "usage: $(basename "$0") <path to SuperCollider.app>"
    exit 1
fi

MACOS_DIR="$APP/Contents/MacOS"

if [ ! -d "$MACOS_DIR" ]; then
    echo "not a directory: $MACOS_DIR"
    exit 1
fi

SCRIPT_NAME='SuperCollider-disable-gpu'
SCRIPT="$MACOS_DIR/$SCRIPT_NAME"

if [ -e "$SCRIPT" ]; then
    echo "$SCRIPT already exists"
    exit 1
fi

SCRIPT_CONTENT='#!/bin/bash
DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
exec "$DIR/SuperCollider" --disable-gpu'

echo "creating $SCRIPT"

echo "$SCRIPT_CONTENT" > "$SCRIPT"

echo "setting execute permissions on $SCRIPT"

chmod a+x "$SCRIPT"

INFO_PLIST="$APP/Contents/Info.plist"

echo "setting CFBundleExecutable = $SCRIPT_NAME in $INFO_PLIST"

defaults write "$INFO_PLIST" CFBundleExecutable "$SCRIPT_NAME"

# `defaults write` saves a binary plist, convert back to xml
plutil -convert xml1 "$INFO_PLIST"

echo 'done'
echo 'rebuilding the LaunchServices database may be needed'
```

only works with the unsigned .app of course.

**Comment by dyfer** [comment]
Is this still an issue with SC 3.12.1 ?

**Comment by zacksettel** [comment]
Running 

SuperCollider 3.12.1

Built from branch 'develop' [cea67fc]



On AppleOS Monterey, on Silicon hardware (mbookPro), no issues to report


Z

**Comment by dyfer** [comment]
> no issues to report

Thanks! In this case let's close this.

**Comment by zacksettel** [comment]
Sweet!
